File: Python/pkg/component/protocol.py
"""
    Package 'component' provides definitions for various objects meeting the definition of a protocol component.
"""

import json
import logging
import random
import time
from constants import Direction
from pkg.component.message import Message
from pkg.component.prototype import Prototype
from pkg.component.device import Device
from pkg.element.segment import Segment
from pkg.element.checksum import Checksum
from pkg.transform.conversion import Conversion
from pkg.transport.simulation import Simulation


"""
    Objects of type Protocol represent a communication protocol consisting of transmit and receive prototypes
	and devices with specific message parameters. Messages are constructed by applying the appropriate prototype
	definition to the device message parameters.

	The Protocol object manages data conversion, checksum handling and transaction simulation.

    Required fields include protocol containing prototype and device definitions.
    Optional fields include prefix, suffix, timeout, source_address, transmission_mode, and checksum_calculation.

"""

class Protocol:

    def __init__(self, protocol: json):
        # Setup logging
        self.logger = logging.getLogger(__name__)
        # Initialize protocol from JSON
        file = None
        # Ensure all required fields are present
        if 'protocol' in protocol:
            file = protocol['protocol']
            # Initialize optional protocol fields
            prefix = ""
            suffix = ""
            timeout = 0
            source_address = 0
            transmission_mode = "hex"
            checksum_calculation = "CRC16"
            # Populate protocol fields if they exist
            if 'prefix' in file:
                prefix = file['prefix']
            if 'suffix' in file:
                suffix = file['suffix']
            if 'timeout' in file:
                timeout = file['timeout']
            if 'source_address' in file:
                source_address = file['source_address']
            if 'transmission_mode' in file:
                transmission_mode = file['transmission_mode']
            if 'checksum_calculation' in file:
                checksum_calculation = file['checksum_calculation']
            # Initialize protocol
            self.__initialize(prefix, suffix, timeout, source_address, transmission_mode, checksum_calculation)
            self.__init_prototypes(file)
            self.__init_devices(file)
        else:
            raise AttributeError("Protocol: protocol object is required")


    def __initialize(self, prefix: str, suffix: str, timeout: int, source_address: int, transmission_mode: str, checksum_calculation: str):        
        self.prefix = prefix
        self.suffix = suffix
        self.timeout = timeout
        self.source_address = source_address
        self.conversion = Conversion(transmission_mode, self.logger)
        self.checksum = Checksum(checksum_calculation, self.conversion, self.logger)
        self.simulation = Simulation(self.conversion, self.logger)
        random.seed(time.time())

    def __init_prototypes(self, file: json):
        if 'prototype' in file:
            self.prototype = []
            self.logger.debug(f"Initializing Prototype: {file["prototype"]}")
            for prototype in file['prototype']:
                self.prototype.append(Prototype(prototype, self.logger))
        else:
            raise AttributeError("Protocol: prototype object is required")
    

    def __init_devices(self, file: json):
        if 'device' in file:
            self.device = []
            self.logger.debug(f"Initializing Device: {file["device"]}")
            for device in file['device']:
                self.device.append(Device(device, self.logger))
        else:
            raise AttributeError("Protocol: device object is required")
    

    def log(self):
        """
        Outputs protocol information.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.logger.debug(f"Protocol: Prefix: {self.prefix}, Suffix: {self.suffix}, Timeout: {self.timeout}, source_address {self.source_address}")

    def run(self):
        """
        Runs the protocol by setting up messages and performing transactions.

        Args:
            None
        Returns:
            None
        Raises:
            None
        """
        self.logger.debug("Protocol: running")
        # Set messages for transmission - currently assumes client (master) role
        self.set_messages_from_prototype(Direction.TX)
        # Perform transactions
        for _ in range(10):
            for device in self.device:
                for msg in device.messages:
                    self.transact(msg, device)
                    

    def set_messages_from_prototype(self, direction: Direction):
        """
        Sets messages for all devices based on prototypes and direction.

        Args:
            direction: The direction of the message (TX or RX).
        Returns:
            None
        Raises:
            None
        """
        for device in self.device:
            for msg in device.messages:
                prototype = self.get_prototype(msg.name)
                if prototype is not None:
                    msg.set_message(self.get_message_from_prototype(prototype, direction, msg, device))
            
    
    def get_message_from_prototype(self, prototype: Prototype, direction: Direction, message: Message, device: Device):
        """
        Applies the prototype definition to the device message parameters to construct the message according to the 
        specified direction, then applies conversion according to the configured transmission mode.

        Args:
            prototype (Prototype): The prototype object.
            direction (Direction): The direction of the message (TX or RX).
            message (Message): The message object.
            device (Device): The device object.
        Returns:
            bytearray: The constructed message bytearray.
        Raises:
            ValueError: If unable to find segments for the specified direction.
            
        """
        msg = bytearray()
        # Get segments based on direction
        segments = prototype.get_segments(direction)
        if segments is None:
            raise ValueError(f"Protocol: unable to find segments for direction: {direction}")
        # Prefix
        msg = self.append_str_to_byte_array(msg, self.prefix)
        # Body
        self.append_segments_to_byte_array(msg, segments, message, device)
        # Suffix
        msg = self.append_str_to_byte_array(msg, self.suffix)
        # Conversion
        return self.conversion.get_converted_message(msg, self.prefix, self.suffix)
    
    # Simulated transaction method
    def transact(self, message: Message, device: Device):
        """
        Simulates a message transaction.

        Args:
            message (Message): The message object.
            device (Device): The device object.
        Returns:
            True if the transaction was successful, False otherwise.
        Raises:
            None
        """
        prototype = self.get_prototype(message.name)
        if prototype is None:
            self.logger.error("Protocol: unable to find prototype for message: %s", message.name)
            return False
        received_msg = self.get_message_from_prototype(prototype, Direction.RX, message, device)
        if received_msg is None:
            self.logger.error("Protocol: no received message for prototype: %s", prototype.name)
            return False
        if self.simulation.simulate_transaction(message.message_byte_array, received_msg):
            if self.checksum.validate_checksum_in_message(received_msg, self.suffix, self.prefix):
                return True
            else:
                self.logger.warning("Protocol: message validation failed for message: %s", message.name)
                return False
        self.logger.warning("Protocol: message transmission failed for message: %s", message.name)
        return False
        
    # Helper method
    def get_prototype(self, name: str):
        """
        Retrieves a prototype by name.

        Args:
            name (str): The name of the prototype.
        Returns:
            The prototype object if found, None otherwise.
        Raises:
            None
        """
        for prototype in self.prototype:
            if prototype.name == name:
                return prototype
        return None
    
    def append_str_to_byte_array(self, msg: bytearray, addition: str):
        """
        Appends a string to the message bytearray.

        Args:
            msg (bytearray): The bytearray to append to.
            addition (str): The string to append.
        Returns:
            bytearray: The updated bytearray.
        Raises:
            None
        """
        if len(addition) > 0:
            for char in addition:
                msg.append(ord(char.upper()))
        return msg        
    
    def append_segments_to_byte_array(self, msg: bytearray, segments: list[Segment], message: Message, device: Device):
        """
        Appends a data representation of each segment to the message bytearray.

        Args:
            msg (bytearray): The bytearray to append to.
            segments (list[Segment]): The list of segments to append.
            message (Message): The message object.
            device (Device): The device object.
        Returns:
            bytearray: The updated bytearray.
        Raises:
            ValueError: If a segment value cannot be found.
        """
        for segment in segments:
            array, bits = self.get_value_array_with_bits(segment, message, device, msg)
            if array is not None:
                for a in array:
                    # Remove '0x' prefix
                    hexadecimal = hex(a)[2:]
                    # Ensure even number of digits
                    if len(hexadecimal) % 2 != 0:
                        hexadecimal = '0' + hexadecimal
                    byte_count = bits / 8 * 2
                    # Pad with leading zeros
                    hexadecimal = hexadecimal.zfill(int(byte_count))
                    # Append bytes to message
                    for i in range(int(byte_count / 2)):                    
                        byte_str = hexadecimal[i*2:i*2+2]
                        byte_val = int(byte_str, 16)
                        msg.append(byte_val)                
            else:
                raise ValueError("Protocol: unable to find value for segment: %s", segment.name)
        return msg

    def get_value_array_with_bits(self, segment: Segment, message: Message, device: Device, msg: bytearray):
        """
        Retrieves the value array and bit size for a given segment.

        Args:
            segment (Segment): The segment object.
            message (Message): The message object.
            device (Device): The device object.
            msg (bytearray): The message bytearray.
        Returns:
            list[int], int: The array of values, and the number of bits.
        Raises:
            None
        """
        array = list[int]()
        if segment.name == "slave_address":
            array.append(device.address)
            return array, segment.bits
        if segment.name == "error_check":
            array.append(self.checksum.calculate_checksum(msg, self.prefix))
            return array, segment.bits
        if segment.name == "byte_count":            
            array.append(int(message.get_data_byte_count()))
            return array, segment.bits
        if segment.name == "data_bytes":            
            b = segment.bits * int(message.get_data_byte_count())
            while b > 0:
                rand = random.randint(0, 255)
                array.append(rand)
                b -= 8
            return array, segment.bits
        # Get value from message based on segment name
        if hasattr(message, segment.name):
            array.append(getattr(message, segment.name))
            return array, segment.bits
        return None, 0
    

File: Go/pkg/component/protocol.go
package component

/*
	Package 'component' provides definitions for various objects meeting the definition of a protocol component.
*/

import (
	"donp/pkg/element"
	"donp/pkg/transform"
	"donp/pkg/transport"
	"fmt"
	"log/slog"
	"math/rand"
	"strconv"
	"strings"
	"time"
)

/*
	Objects of type Protocol represent a communication protocol consisting of transmit and receive prototypes
	and devices with specific message parameters. Messages are constructed by applying the appropriate prototype
	definition to the device message parameters.

	The Protocol object manages data conversion, checksum handling and transaction simulation.

	Required fields include protocol containing prototype and device definitions.
    Optional fields include prefix, suffix, timeout, source_address, transmission_mode, and checksum_calculation.
*/

const (
	// DirectionTransmit indicates the transmit direction.
	DirectionTX = "transmit"
	// DirectionReceive indicates the receive direction.
	DirectionRX = "receive"
)

type Direction string

// Protocol represents a communication protocol consisting of multiple prototypes.
type Protocol struct {
	Prefix              string                `json:"prefix"`
	Suffix              string                `json:"suffix"`
	Timeout             uint                  `json:"timeout"`
	SourceAddress       uint                  `json:"source_address"`
	TransmissionMode    string                `json:"transmission_mode"`
	ChecksumCalculation string                `json:"checksum_calculation"`
	Prototype           []*Prototype          `json:"prototype"`
	Devices             []*Device             `json:"device"`
	RandomGenerator     *rand.Rand            `json:"-"`
	Conversion          *transform.Conversion `json:"-"`
	Checksum            *element.Checksum     `json:"-"`
	Simulation          *transport.Simulation `json:"-"`
}

// NewProtocol creates a new instance of Protocol.
func NewProtocol(dataMap map[string]interface{}) *Protocol {
	// Log level set to Info for protocol operations
	slog.SetLogLoggerLevel(slog.LevelInfo)
	// Ensure protocol field exists
	if dataMap["protocol"] == nil {
		slog.Error("Protocol: protocol object is required")
		return nil
	}
	// Create protocol instance
	var protocol Protocol
	protocolMap := dataMap["protocol"].(map[string]interface{})
	// Ensure prototype and device fields exist
	if protocolMap["prototype"] == nil {
		slog.Error("Protocol: prototype object is required")
		return nil
	}
	if protocolMap["device"] == nil {
		slog.Error("Protocol: device object is required")
		return nil
	}
	// Initialize optional fields with defaults
	protocol.Prefix = ""
	protocol.Suffix = ""
	protocol.Timeout = 0
	protocol.SourceAddress = 0
	protocol.TransmissionMode = "hex"
	protocol.ChecksumCalculation = "CRC16"
	// Populate optional fields
	if prefix, ok := protocolMap["prefix"].(string); ok {
		protocol.Prefix = prefix
	}
	if suffix, ok := protocolMap["suffix"].(string); ok {
		protocol.Suffix = suffix
	}
	if timeout, ok := protocolMap["timeout"].(float64); ok {
		protocol.Timeout = uint(timeout)
	}
	if sourceAddress, ok := protocolMap["source_address"].(float64); ok {
		protocol.SourceAddress = uint(sourceAddress)
	}
	if transmissionMode, ok := protocolMap["transmission_mode"].(string); ok {
		protocol.TransmissionMode = transmissionMode
	} else {
		protocol.TransmissionMode = "hex"
	}
	if checksumCalculation, ok := protocolMap["checksum_calculation"].(string); ok {
		protocol.ChecksumCalculation = checksumCalculation
	} else {
		protocol.ChecksumCalculation = "CRC16"
	}
	protocol.Conversion = transform.NewConversion(protocol.TransmissionMode)
	protocol.Checksum = element.NewChecksum(protocol.ChecksumCalculation, *protocol.Conversion)
	protocol.Simulation = transport.NewSimulation(*protocol.Conversion)
	// Make prototype
	for _, protoData := range protocolMap["prototype"].([]interface{}) {
		protoMap := protoData.(map[string]interface{})
		protoType, err := NewPrototype(protoMap)
		if err != nil {
			slog.Error("Protocol: Failed to create prototype: " + err.Error())
			continue
		}
		protocol.Prototype = append(protocol.Prototype, protoType)
	}
	if len(protocol.Prototype) == 0 {
		slog.Error("Protocol: prototype object is required")
		return nil
	}
	// Make devices
	for _, devData := range protocolMap["device"].([]interface{}) {
		devMap := devData.(map[string]interface{})
		protocol.Devices = append(protocol.Devices, NewDevice(devMap))
	}
	if len(protocol.Devices) == 0 {
		slog.Error("Protocol: device object is required")
		return nil
	}
	protocol.RandomGenerator = rand.New(rand.NewSource(time.Now().UnixNano()))
	return &protocol
}

// Log outputs protocol information,
func (p *Protocol) Log() {
	slog.Debug("Protocol: Prefix = " + p.Prefix + ", Suffix = " + p.Suffix + ", Timeout = " + strconv.Itoa(int(p.Timeout)) + " ms, Source Address = " + strconv.Itoa(int(p.SourceAddress)))
}

// Run runs the protocol by setting up messages and performing transactions.
func (p *Protocol) Run() {
	slog.Debug("Protocol: running")
	// Set messages for transmission - currently assumes client (master) role
	p.setMessagesFromPrototype(DirectionTX)
	// Perform transactions
	for i := 0; i < 10; i++ {
		for _, device := range p.Devices {
			for _, message := range device.Messages {
				p.transact(message, device)
			}
		}
	}
}

// setMessagesFromPrototype sets messages for all devices based on prototypes and direction.
func (p *Protocol) setMessagesFromPrototype(direction Direction) {
	for _, device := range p.Devices {
		for _, msg := range device.Messages {
			protoType := p.getPrototype(msg.Name)
			if protoType != nil {
				msg.SetMessage(p.getMessageFromPrototype(protoType, direction, msg, device))
			}
		}
	}
}

// getMessageFromPrototype applies the prototype definition to the device message parameters
// to construct the message according to the specified direction, then applies conversion
// according to the configured transmission mode.
func (p *Protocol) getMessageFromPrototype(proto *Prototype, direction Direction, message *Message, device *Device) []byte {
	var msg []byte
	// Get segments based on direction
	segments := proto.GetSegments(direction)
	if segments == nil {
		slog.Error("Protocol: unable to find segments for direction: " + string(direction))
		return nil
	}
	// Prefix
	msg = p.appendStrToByteArray(msg, p.Prefix)
	// Body
	msg = p.appendSegmentsToByteArray(msg, segments, *message, device)
	// Suffix
	msg = p.appendStrToByteArray(msg, p.Suffix)
	// Conversion
	return p.Conversion.GetConvertedMessage(msg, p.Prefix, p.Suffix)
}

// Simulated transaction method
func (p *Protocol) transact(message *Message, device *Device) bool {
	prototype := p.getPrototype(message.Name)
	if prototype == nil {
		slog.Error("Protocol: no prototype found for message: " + message.Name)
		return false
	}
	receivedMsg := p.getMessageFromPrototype(prototype, DirectionRX, message, device)
	if receivedMsg == nil {
		slog.Warn("Protocol: no received message for prototype: " + prototype.Name)
		return false
	}
	if p.Simulation.SimulateTransaction(message.Message, receivedMsg) {
		if p.Checksum.ValidateChecksumInMessage(receivedMsg, p.Prefix, p.Suffix) {
			return true
		} else {
			slog.Warn("Protocol: message validation failed for message: " + message.Name)
			return false
		}
	}
	slog.Warn("Protocol: message transmission failed for message: " + message.Name)
	return false
}

// Helper methods
func (p *Protocol) getPrototype(name string) *Prototype {
	for _, proto := range p.Prototype {
		if proto.Name == name {
			return proto
		}
	}
	return nil
}

// appendStrToByteArray appends a string to the byte array.
func (p *Protocol) appendStrToByteArray(msg []byte, addition string) []byte {
	if len(addition) > 0 {
		addition = strings.ToUpper(addition)
		add := []byte(addition)
		msg = append(msg, add...)
	}
	return msg
}

// appendSegmentsToByteArray appends a data representation of each segment to the message byte array.
func (p *Protocol) appendSegmentsToByteArray(msg []byte, segments []*element.Segment, message Message, device *Device) []byte {
	for _, segment := range segments {
		array, bits := p.getValueArrayWithBits(segment, &message, device, msg)
		if array != nil {
			for a := range array {
				// Remove 0x prefix
				hexadecimal := fmt.Sprintf("%02X", array[a])
				// Ensure even number of digits
				if len(hexadecimal)%2 != 0 {
					hexadecimal = "0" + hexadecimal
				}
				byteCount := bits / 8 * 2
				// Pad with leading zero if necessary
				for len(hexadecimal) < int(byteCount) {
					hexadecimal = "0" + hexadecimal
				}
				// Append bytes to message
				for i := range int(byteCount) / 2 {
					byte_str := hexadecimal[i*2 : i*2+2]
					byte_val, _ := strconv.ParseInt(byte_str, 16, 0)
					msg = append(msg, byte(byte_val))
				}
			}
		} else {
			slog.Error("Protocol: unable to find value for segment: " + segment.Name)
		}
	}
	return msg
}

// getValueArrayWithBits retrieves the value array and bit size for a given segment.
func (p *Protocol) getValueArrayWithBits(segment *element.Segment, message *Message, device *Device, msg []byte) ([]int, uint) {
	var array []int
	if segment.Name == "slave_address" {
		array = append(array, device.Address)
		return array, uint(segment.Bits)
	}
	if segment.Name == "error_check" {
		array = append(array, p.Checksum.CalculateChecksum(msg, p.Prefix))
		return array, uint(segment.Bits)
	}
	if segment.Name == "byte_count" {
		array = append(array, int(message.getDataByteCount()))
		return array, uint(segment.Bits)
	}
	if segment.Name == "data_bytes" {
		b := int(segment.Bits) * int(message.getDataByteCount())
		for b > 0 {
			rand := p.RandomGenerator.Intn(256)
			array = append(array, rand)
			b -= 8
		}
		return array, uint(segment.Bits)
	}
	if message.MsgDictionary[segment.Name] != nil {
		array = append(array, int(message.MsgDictionary[segment.Name].(float64)))
		return array, uint(segment.Bits)
	}
	return nil, 0
}
